static/js/input.js

```diff
@@ -442,6 +447,41 @@ function guessColumnType(arr) {
 }
 
 
+/**
+ * Calculates the entropy of an array
+ * @function calcEntropy
+ * @param {Array} arr - array of values
+ * @returns {Number} entropy - entropy of array
+ */
+function calcEntropy(arr, match) {
+  const unique = [...new Set(arr)];
+  const base = unique.length;
+  let probs = [];
+
+  // calculates probability of occurence of search value
+  function calcProb(match) {
+    const n = arr.length;
+
+    let count = 0;
+    for (let i = 0; i < n; i++) {
+      if (arr[i] === match) {
+        count++;
+      }
+    }
+
+    return count / n;
+  }
+
+  for (let val of unique) probs.push(calcProb(val));
+
+  const reducer = (acculumator, current) => acculumator + (current + (Math.log(current) / Math.log(base)));
+  const entropy = -1 * probs.reduce(reducer);
```

Comment:
This is a briliant use of the new JavaScript syntax! Very elegant!

However, syntactic elegance does not always translate into performance. In JavaScript, actually, most of the fancy ways of looping an array has worse performance compared with the old-style `for (i = 0; ...)` loop. Here is an [article](https://leanylabs.com/blog/js-forEach-map-reduce-vs-for-for_of/) showing the benchmarks. There are many other such articles.

In this case, performance is favorable because there may be multiple thousands of elements in the array. Therefore I would recommend using the old-style `for` loop, as I did in many other places in the code where the length of code can be reduced by `for ... of`, `forEach`, `reduce` and other tricks but where performance is vital.

Two simple rules regarding algorithm optimization:

1. Calling a function is expensive. The `reducer` you implemented is repeatedly calling a function (after `=>`). Therefore it has a non-negligible overhead.

2. Accessing an array element by its index is constant (_O_(1)). Therefore when you do `for (i = 0; i < n; i++) { arr[i] bruh bruh bruh }` the performance is maximized.

Reply:
What is `base` here? I guess it is 2 (according to the definition of Shannon entropy). In JavaScript there is a [`Math.log2`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/log2) function.